### PR TITLE
Add serial_print

### PR DESCRIPTION
**Pull request by bigdinotech**
-add serial_print library for printing into Serial CDC-ACM port using
the quark core

### CODE REVIEW

**Comment by bigdinotech** [comment]
@SidLeung @eriknyquist ready for review

File: Quark_Sketch/src/arduino/serial_print.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+#if defined(CONFIG_STDOUT_CONSOLE)
+#include <stdio.h>
+#define PRINT           printf
+#else
+#include <misc/printk.h>
+#define PRINT           printk
+#endif
+
+#define Rx_BUFF curie_shared_data->cdc_acm_shared_rx_buffer.data
```
Comment (eriknyquist) [inline Quark_Sketch/src/arduino/serial_print.c]: I remember seeing these same defines in another file. Can they go in a common file? "curie_shared_mem.h" maybe?

File: Quark_Sketch/src/arduino/serial_print.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#ifndef SERIAL_PRINT_H
+#define SERIAL_PRINT_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Comment (eriknyquist) [inline Quark_Sketch/src/arduino/serial_print.h:8]: Indentation after opening brace?

File: Quark_Sketch/src/arduino/serial_print.c [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	{
+		digits++;
+		n /= 10;
+	}
+	
+	sprintf(buf, "%d", val);
+	serial_print(buf);
+}
+
+int serial_write(const uint8_t uc_data)
```
Comment (eriknyquist) [inline Quark_Sketch/src/arduino/serial_print.c]: It would be nice to provide a `serial_writec` for single char and `serial_writes` for a string. I know they can implement `serial_writes` themselves, but then they will use your `serial_write` for each character and the initial check on line 66 will happen for every character. We could eliminate that by implementing a `writes` function that just checks once before sending all the chars.

File: Quark_Sketch/src/arduino/serial_print.c [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+{
+	if(!curie_shared_data->cdc_acm_buffers_obj.device_open)
+		return 0;
+	else
+		return (uint16_t)(SBS + Rx_HEAD - Rx_TAIL)%SBS;
+}
+
+uint16_t serial_availableForWrite()
+{
+	if(!curie_shared_data->cdc_acm_buffers_obj.host_open || !curie_shared_data->cdc_acm_buffers_obj.device_open)
```
Comment (eriknyquist) [inline Quark_Sketch/src/arduino/serial_print.c]: This check (line 87) seems to happen a few times. Would a function be appropriate? `buf_is_open()` or something?

File: Quark_Sketch/src/arduino/serial_print.c [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	int tail = Tx_TAIL;
+
+	if(head >= tail)
+		return SBS - head + tail - 1;
+	return tail - head - 1;
+}
+
+int serial_peek()
+{
+	if(!curie_shared_data->cdc_acm_buffers_obj.host_open || (Rx_HEAD == Rx_TAIL))
```
Comment (eriknyquist) [inline Quark_Sketch/src/arduino/serial_print.c]: Again, maybe a function is appropriate? something like `buf_is_empty()`?

File: Quark_Sketch/prj.conf [hunk 6]
```diff
@@ -18,5 +18,6 @@ CONFIG_GPIO_QMSI_AON=y
 CONFIG_GPIO_QMSI_AON_NAME="GPIO_AON_0"
 CONFIG_GPIO_QMSI_AON_PRI=2
 CONFIG_PWM=y
+CONFIG_PWM_QMSI_NUM_PORTS=4
```
Comment (eriknyquist) [inline Quark_Sketch/prj.conf]: Why is this needed for serial_print functions?

Comment (bigdinotech) [inline Quark_Sketch/prj.conf]: this actually fixes the issue with pwm only working on pin3.
I will remove and create a separate PR

**Comment by bigdinotech** [comment]
@eriknyquist made the appropriate changes. also added serial_printFloat

**Comment by eriknyquist** [comment]
Thanks @bigdinotech, looks good +1

**Comment by bigdinotech** [comment]
@calvinatintel please merge
